src/sabre_swap/mod.rs

```diff
@@ -0,0 +1,147 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#![allow(clippy::too_many_arguments)]
+
+pub mod edge_list;
+pub mod qubits_decay;
+pub mod swap_scores;
+
+use ndarray::prelude::*;
+use numpy::IntoPyArray;
+use numpy::PyReadonlyArray2;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+use rayon::prelude::*;
+
+use crate::nlayout::NLayout;
+
+use edge_list::EdgeList;
+use qubits_decay::QubitsDecay;
+use swap_scores::SwapScores;
+
+const EXTENDED_SET_WEIGHT: f64 = 0.5; // Weight of lookahead window compared to front_layer.
+
+#[pyclass]
+pub enum Heuristic {
+    Basic,
+    Lookahead,
+    Decay,
+}
+
+#[pyfunction]
+pub fn sabre_score_heuristic(
+    py: Python,
+    layer: EdgeList,
+    layout: &NLayout,
+    swap_scores: &mut SwapScores,
+    extended_set: EdgeList,
+    distance_matrix: PyReadonlyArray2<f64>,
+    qubits_decay: QubitsDecay,
+    heuristic: &Heuristic,
+) -> PyObject {
+    let dist = distance_matrix.as_array();
+    swap_scores
+        .scores
+        .iter_mut()
+        .for_each(|(swap_qubits, score)| {
+            let mut trial_layout = layout.clone();
+            trial_layout.swap_logical(swap_qubits[0], swap_qubits[1]);
+            *score = score_heuristic(
+                heuristic,
+                &layer.edges,
+                &extended_set.edges,
+                &trial_layout,
+                swap_qubits,
+                &dist,
+                &qubits_decay.decay,
+            );
+        });
+    let min_score = swap_scores
+        .scores
+        .values()
+        .min_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap();
```

Comment:
Yep, exactly. Since float types can't compare NaN Rust only implements the [`PartialOrd`](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) trait for float types. But the `min()` method requires the `Ord` trait https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.min and doesn't know how to deal with a `None` response from a comparison. This will technically panic if any value is NaN but that didn't seem very likely so I was ok with that. 

That being said I have removed this in my local copy of the branch (which I'm running benchmarks on now) as I removed `SwapScores` and just iterate directly over the swap candidates and build the `best_swaps` list directly so we don't need to iterate over everything again.